lib/widgets/channel_list.dart

```diff
@@ -100,6 +103,82 @@ class ChannelItem extends StatelessWidget {
+
+  final ZulipStream stream;
+  final BuildContext channelItemContext;
+
+  @override
+  State<_ChannelItemSubscribeButton> createState() => _ChannelItemSubscribeButtonState();
+}
+
+class _ChannelItemSubscribeButtonState extends State<_ChannelItemSubscribeButton> {
+  bool _isLoading = false;
+
+  void _setIsLoading(bool value) {
+    if (!mounted) return;
+    setState(() => _isLoading = value);
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return IconButton(
+      icon: const Icon(Icons.add),
+      onPressed: _isLoading ? null : () async {
+        _setIsLoading(true);
+        await _subscribeToChannel(context, widget.stream);
+        _setIsLoading(false);
+      });
+  }
+
+  Future<void> _subscribeToChannel(BuildContext context, ZulipStream stream) async {
+    final store = PerAccountStoreWidget.of(context);
+    final connection = store.connection;
+    final zulipLocalizations = ZulipLocalizations.of(context);
+    try {
+      final res = await subscribeToChannels(connection, [stream]);
+      if (!context.mounted) return;
+      if (_subscriptionResultContains(store, res.subscribed, stream.name)) {
+        _showSnackbarWithChannelTitle(context, zulipLocalizations.messageSubscribedToChannel, stream);
+      } else if (_subscriptionResultContains(store, res.alreadySubscribed, stream.name)) {
+        _showSnackbarWithChannelTitle(context, zulipLocalizations.messageAlreadySubscribedToChannel, stream);
+      } else {
+        _showSnackbarWithChannelTitle(context, zulipLocalizations.errorFailedToSubscribeToChannel, stream);
+      }
+    } catch (e) {
+      if (!context.mounted) return;
+      final zulipLocalizations = ZulipLocalizations.of(context);
+      _showSnackbarWithChannelTitle(context, zulipLocalizations.errorFailedToSubscribeToChannel, stream);
+    }
+  }
+
+  bool _subscriptionResultContains(PerAccountStore store, Map<String, List<String>> emailOrIdSubs, String subscription) {
+    // TODO (server-10) Before the user keys were Zulip API email addresses, not user ID.
+    final expectedEmail = store.users[store.selfUserId]?.email;
+    final foundByEmail = emailOrIdSubs[expectedEmail]?.contains(subscription);
+    final foundById = emailOrIdSubs[store.selfUserId.toString()]?.contains(subscription);
+    return foundByEmail ?? foundById ?? false;
+  }
+
+  void _showSnackbarWithChannelTitle(BuildContext context, String message, ZulipStream stream) {
+    ScaffoldMessenger.of(context).showSnackBar(
+      SnackBar(
+        behavior: SnackBarBehavior.floating,
+        content: Text.rich(TextSpan(text: message, children: [
+          WidgetSpan(child: Padding(
+            padding: const EdgeInsets.only(bottom: 2),
+            child: Icon(
+              size: 12,
+              iconDataForStream(stream),
+              color: Theme.of(context).scaffoldBackgroundColor))),
+          TextSpan(text: ' ${stream.name}.', style: weightVariableTextStyle(context, wght: 700)),
```

Comment:
Oh, I see. Regarding the earlier comment on the translated string, this wouldn't work for all languages because the order might be different. We can just keep it simple and style the message the same way.

Reply:
Yeah, to make this work properly we'd want to end up writing something like:
```dart
Text.rich(zulipLocalizations.errorFailedToSubscribeToChannel(
  channel: TextSpan(children: [
    WidgetSpan(/* … */),
    TextSpan(text: ' ${channel.name}', /* … bold … */),
  ]))
```

Then the string for translation would look like
"Failed to subscribe to {channel}."

and the generated implementation for `errorFailedToSubscribeToChannel`, for English, would look like
```dart
  TextSpan errorFailedToSubscribeToChannel({required InlineSpan channel}) {
    return TextSpan(children: [
      const TextSpan(text: 'Failed to subscribe to '),
      channel,
      const TextSpan(text: '.'),
    ]);
  }
```

Flutter's tooling for l10n doesn't currently have that feature, though. There's an open issue for it:
* https://github.com/flutter/flutter/issues/147974

Reply:
One workaround would be to make the string for translation be:
"Failed to subscribe to channel: "

with a colon. Then we can show the translated version of that, followed by the fancy formatted spans, and no period.

Basically by making the message be less in the format of normal prose, and more in the format of a form like something printed out by a computer system, we ensure the translator can make something where the channel name (and attached icon) naturally goes at the end. In different languages the place that the object of a verb like "subscribe" goes, or that the whole relative clause "to subscribe …" goes in a sentence with "failed", can vary and may not come at the end… but every language (that anyone would translate our UI into) is used to dealing with forms and bureaucracies and computers, and has ways to express a line in a computery bureaucratic form.